### PR TITLE
Disable resync on SDK Server

### LINKED ISSUE #3377: Disable resync on SDK Server (?)
**Is your feature request related to a problem? Please describe.**

An idea I had the other day, each game server resyncs every 30s (like the controller), but _does it need to_?

If it doesn't need to, that could potentially take some load off the KCP.

This would mean that the `SDK.Watch()` command wouldn't get pinged every 30s (which may be a good or bad thing), regardless of whether there is a change or not.

**Describe the solution you'd like**

Turn off resync here (assuming you can? I can't seem to find docs on that either):

https://github.com/googleforgames/agones/blob/963fb156d492293e41c5e68d01ea8d184806446a/pkg/sdkserver/sdkserver.go#L135-L135

Or set it to some ridiculously high value.

**Describe alternatives you've considered**

Leave it alone. It might just be fine.

I think research might be required to determine if resync is really needed or not (what issues does it paper over with the KCP connection and interaction).

**Additional context**

* [Google Group: What's the right resync period (2017)](https://groups.google.com/g/kubernetes-sig-api-machinery/c/PbSCXdLDno0)

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #3377 

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/sdkserver/sdkserver.go [hunk 1]
```diff
@@ -144,7 +144,7 @@ func NewSDKServer(gameServerName, namespace string, kubeClient kubernetes.Interf
 		s1 := fields.OneTermEqualSelector("metadata.name", gameServerName)
 		opts.FieldSelector = s1.String()
 	}
-	factory := externalversions.NewSharedInformerFactoryWithOptions(agonesClient, 30*time.Second, externalversions.WithNamespace(namespace), externalversions.WithTweakListOptions(tweakListOptions))
+	factory := externalversions.NewSharedInformerFactoryWithOptions(agonesClient, 0, externalversions.WithNamespace(namespace), externalversions.WithTweakListOptions(tweakListOptions))
```
Comment (markmandel) [inline pkg/sdkserver/sdkserver.go:147]: I've always wondered if 0 was "never" - did you find  a doc on that? I never managed to actually find an answer either way.

Comment (Kalaiselvi84) [reply pkg/sdkserver/sdkserver.go:147]: https://github.com/kyverno/kyverno/issues/2792
https://github.com/kyverno/kyverno/pull/2806/files

They gave it a shot with 0, but then they decided they didn't need it anymore and closed it.

Comment (Kalaiselvi84) [reply pkg/sdkserver/sdkserver.go:147]: https://firehydrant.com/blog/stay-informed-with-kubernetes-informers/

Comment (markmandel) [reply pkg/sdkserver/sdkserver.go:147]: Took me a bit, but I found the code! You are correct - 0 is _off_.

https://github.com/kubernetes/client-go/blob/84a6fe7e4032ae1b8bc03b5208e771c5f7103549/tools/cache/shared_informer.go#L283-L286

Comment (Kalaiselvi84) [reply pkg/sdkserver/sdkserver.go:147]: Shall we keep this PR and close the other PR?

Comment (markmandel) [reply pkg/sdkserver/sdkserver.go:147]: Let's keep both running - curious to see what the e2e results are.

I'm leaning more towards this one, but let's see what happens.

Comment (Kalaiselvi84) [reply pkg/sdkserver/sdkserver.go:147]: quick question:
Informers keep an eye on what's happening in Kubernetes resources and keeping the local cache up to date with changes happening in the cluster, right? 
Disable resync will be helpful when there are no resource changes, but when there are changes in the resources how do informers function?

Comment (markmandel) [reply pkg/sdkserver/sdkserver.go:147]: > Informers keep an eye on what's happening in Kubernetes resources and keeping the local cache up to date with changes happening in the cluster, right?

100% nailed it.

> Disable resync will be helpful when there are no resource changes, but when there are changes in the resources how do informers function?

Informers use a Kuberntes watch operations to keep up to date with changes that have happened in the K8s cluster.

See: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#:~:text=Watch%3A%20Watch%20will%20stream%20results%20for%20an%20object(s)%20as%20it%20is%20updated.%20Similar%20to%20a%20callback%2C%20watch%20is%20used%20to%20respond%20to%20resource%20changes.

What resync does it forces a full list of all the K8s resources rather than just pushing changes to the Informer.

Next steps:
Let's close the other PR, since this pass e2e tests - and let's move this change behind a Feature Flag, so we can test it with it enabled and disabled, and see the performance impact!

We can do this all in this PR if that's easiest for you.

So we'll need a new feature flag in here:

Comment (markmandel) [reply pkg/sdkserver/sdkserver.go:147]: Updated above comment - hit submit too early.

Comment (Kalaiselvi84) [reply pkg/sdkserver/sdkserver.go:147]: Closed the other PR.
Followed these steps: https://github.com/googleforgames/agones/blob/25e0c4a3e85688d181b5f155d1ef66b576d839bc/pkg/util/runtime/features.go#L77-L83

I am confused with true and false [here](https://github.com/googleforgames/agones/blob/25e0c4a3e85688d181b5f155d1ef66b576d839bc/pkg/util/runtime/features.go#L82)😅. Please review and share your thoughts👍🏻

File: install/helm/agones/defaultfeaturegates.yaml [hunk 2]
```diff
@@ -20,9 +20,10 @@ ResetMetricsOnDelete: true
 SplitControllerAndExtensions: true
 
 # Alpha features
+DisableResyncOnSDKServer: false
+FleetAllocationOverflow: false
```
Comment (markmandel) [inline install/helm/agones/defaultfeaturegates.yaml:24]: Please don't change the order or things - makes it hard to follow the changes.

File: build/Makefile [hunk 3]
```diff
@@ -64,7 +64,7 @@ KIND_CONTAINER_NAME=$(KIND_PROFILE)-control-plane
 GS_TEST_IMAGE ?= us-docker.pkg.dev/agones-images/examples/simple-game-server:0.22
 
 # Enable all alpha feature gates. Keep in sync with `false` (alpha) entries in pkg/util/runtime/features.go:featureDefaults
-ALPHA_FEATURE_GATES ?= "PlayerAllocationFilter=true&PlayerTracking=true&CountsAndLists=true&FleetAllocationOverflow=true&Example=true"
+ALPHA_FEATURE_GATES ?= "CountsAndLists=true&DisableResyncOnSDKServer=true&Example=true&FleetAllocationOverflow=true&PlayerAllocationFilter=true&PlayerTracking=true"
```
Comment (markmandel) [inline build/Makefile:67]: Can we please keep the original ordering, and append `DisableResyncOnSDKServer` on the end (or before `Example`) - makes things much easier to review and ensure nothing changed that it shouldn't.

File: cloudbuild.yaml [hunk 4]
```diff
@@ -296,10 +296,10 @@ steps:
             region=${versionsAndRegions[$version]}
             if [ $cloudProduct = generic ]
             then
-              featureWithGate="PlayerAllocationFilter=true&PlayerTracking=true&ResetMetricsOnDelete=false&PodHostname=false&SplitControllerAndExtensions=false&FleetAllocationOverflow=true&CountsAndLists=true&Example=true"
+              featureWithGate="CountsAndLists=true&DisableResyncOnSDKServer=true&Example=true&FleetAllocationOverflow=true&PlayerAllocationFilter=true&PlayerTracking=true&ResetMetricsOnDelete=false&PodHostname=false&SplitControllerAndExtensions=false"
```
Comment (markmandel) [inline cloudbuild.yaml:299]: Same here - retain original ordering, and append to end.

File: site/content/en/docs/Guides/feature-stages.md [hunk 5]
```diff
@@ -34,7 +35,20 @@ The current set of `alpha` and `beta` feature gates:
...
 | [Allocated GameServers are notified on relevant Fleet Updates][fleet-updates]                                         | `FleetAllocationOverflow`      | Disabled | `Alpha` | 1.32.0 |
 | Example Gate (not in use)                                                                                             | `Example`                      | Disabled | None    | 0.13.0 |
+{{% /feature %}}
 
+{{% feature publishVersion="1.37.0" %}}
+| Feature Name                                                                                                          | Gate                           | Default  | Stage   | Since  |
+|-----------------------------------------------------------------------------------------------------------------------|--------------------------------|----------|---------|--------|
+| [DisableResyncOnSDKServer](https://github.com/googleforgames/agones/issues/3377)                                                                                             | `DisableResyncOnSDKServer`                      | Disabled | `Alpha`    | 1.37.0 |
```
Comment (markmandel) [inline site/content/en/docs/Guides/feature-stages.md:43]: Again, please retain the original ordering - add the new one before `Example`.

File: pkg/sdkserver/sdkserver.go (hunk 1 above)
Comment (markmandel) [inline pkg/sdkserver/sdkserver.go:147]: Next step: wrap this in a feature gate method with the new feature gate, so that the when true, resync is 0, and when disabled (currently the default), it's still 30s.

https://github.com/googleforgames/agones/blob/25e0c4a3e85688d181b5f155d1ef66b576d839bc/pkg/util/runtime/features.go#L204

Is the function.

For example:

https://github.com/googleforgames/agones/blob/065052a9beead20bbc2d7bf9a18a7906bc73a882/pkg/sdkserver/sdk.go#L78-L87

**Review by markmandel — COMMENTED** [review]
On first pass - the approach looks spot on. Just need to revert the reordering, and add the conditional step (comments inline).

**Comment by markmandel** [comment]
I did just see the line in the instructions that says:
> Ensure that the features in each file are organized alphabetically.

Yeah - ignore that for now 😄

**Comment by Kalaiselvi84** [comment]
> Yeah - ignore that for now 😄

My apologies, I overlooked that. Retained the order now.

File: pkg/sdkserver/sdkserver.go [hunk 6]
```diff
@@ -138,13 +138,17 @@ type SDKServer struct {
 func NewSDKServer(gameServerName, namespace string, kubeClient kubernetes.Interface,
 	agonesClient versioned.Interface) (*SDKServer, error) {
 	mux := http.NewServeMux()
+	resync := 30 * time.Second
+	if runtime.FeatureEnabled(runtime.FeatureDisableResyncOnSDKServer) {
```
Comment (markmandel) [inline pkg/sdkserver/sdkserver.go:142]: 👍🏻

**Review by markmandel — APPROVED** [review]
Nice!
